tests/unattended/unit/framework/bach.sh

```diff
@@ -0,0 +1,642 @@
+# -*- mode: sh -*-
+# Bach Testing Framework, https://bach.sh
+# Copyright (C) 2019  Chai Feng <chaifeng@chaifeng.com>
+#
+# Bach Testing Framework is dual licensed under:
+# - GNU General Public License v3.0
+# - Mozilla Public License 2.0
+set -euo pipefail
+shopt -s expand_aliases
+
+export BACH_COLOR="${BACH_COLOR:-auto}"
+export PS4='+(${BASH_SOURCE##*/}:${LINENO}): ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+
+BACH_OS_NAME="$(uname)"
+declare -gxr BACH_OS_NAME
+
+declare -gxa bach_origin_paths=()
+while builtin read -r -d: folder; do
+    bach_origin_paths+=("$folder")
+done <<< "${PATH}:"
+
+function @out() {
+    if [[ "${1:-}" == "-" || ! -t 0 ]]; then
+        [[ "${1:-}" == "-" ]] && shift
+        while IFS=$'\n' read -r line; do
+            builtin printf "%s\n" "${*}$line"
+        done
+    elif [[ "$#" -gt 0 ]]; then
+        builtin printf "%s\n" "$*"
+    else
+        builtin printf "\n"
+    fi
+} 8>/dev/null
+export -f @out
+
+function @err() {
+    @out "$@"
+} >&2
+export -f @err
+
+function @die() {
+    @out "$@"
+    exit 1
+} >&2
+export -f @die
+
+if [[ -z "${BASH_VERSION:-}" ]] || [[ "${BASH_VERSINFO[0]}" -lt 4 ]]; then
+    @die "Bach Testing Framework only support Bash v4+!"
+fi
+
+if [[ "${BACH_DEBUG:-}" != true ]]; then
+    function @debug() {
+        :
+    }
+else
+    exec 8>&2
+    function @debug() {
+        builtin printf '[DEBUG] %s\n' "$*"
+    } >&8
+fi
+export -f @debug
+
+function bach-real-path() {
+    declare folder name="$1"
+    declare altname="${name#*|}"
+    name="${name%|*}"
+    for folder in "${bach_origin_paths[@]}"; do
+        if [[ -x "$folder/$name" ]]; then
+           builtin echo "$folder/$name"
+           return 0
+        elif [[ "$name" != "$altname" && -x "$folder/$altname" ]]; then
```

Comment:
Same

Reply:
`bach.sh` is the external framework. We will not modify it.